src/request/codec.rs

```diff
@@ -0,0 +1,343 @@
+        }
+
+        Ok(())
+    }
+
+    fn decode_kvs(&self, kvs: &mut [kvrpcpb::KvPair]) -> Result<()> {
+        self.decode_pairs(kvs)
+    }
+
+    fn decode_regions(&self, regions: &mut [Region]) -> Result<()> {
+        for region in regions.iter_mut() {
+            self.decode_region(region)?;
+        }
+        Ok(())
+    }
+
+    fn decode_region_error(&self, err: &mut errorpb::Error) -> Result<()> {
+        if err.has_epoch_not_match() {
+            self.decode_regions(err.mut_epoch_not_match().mut_current_regions())?;
+        }
+        Ok(())
+    }
+}
+
+impl<T: RequestCodec> RequestCodecExt for T {}
+
+pub trait RawCodec: RequestCodec {}
+
+pub trait TxnCodec: RequestCodec {}
+
+#[derive(Clone, Copy, Default, PartialEq, Eq)]
+pub struct KeySpaceId([u8; 3]);
+
+impl Deref for KeySpaceId {
+    type Target = [u8; 3];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for KeySpaceId {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+pub trait Mode: Clone + Copy + Sync + Send + 'static {
+    const PREFIX: u8;
+    const MIN_KEY: &'static [u8] = &[Self::PREFIX, 0, 0, 0];
+    const MAX_KEY: &'static [u8] = &[Self::PREFIX + 1, 0, 0, 0];
+}
+
+#[derive(Clone, Copy)]
+pub struct RawMode;
+
+#[derive(Clone, Copy)]
+pub struct TxnMode;
+
+impl Mode for RawMode {
+    const PREFIX: u8 = b'r';
+}
+
+impl Mode for TxnMode {
+    const PREFIX: u8 = b't';
+}
+
+#[derive(Clone)]
+pub struct ApiV1<M: Mode> {
+    _phantom: PhantomData<M>,
+}
+
+impl<M: Mode> Default for ApiV1<M> {
+    fn default() -> Self {
+        ApiV1 {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl RequestCodec for ApiV1<RawMode> {}
+
+impl RawCodec for ApiV1<RawMode> {}
+
+impl RequestCodec for ApiV1<TxnMode> {
+    fn encode_pd_query(&self, key: Vec<u8>) -> Vec<u8> {
+        Key::from(key).to_encoded().into()
+    }
+
+    fn decode_region(&self, region: &mut Region) -> Result<()> {
+        decode_bytes_in_place(region.mut_start_key(), false)?;
+        decode_bytes_in_place(region.mut_end_key(), false)?;
+
+        Ok(())
+    }
+}
+
+impl TxnCodec for ApiV1<TxnMode> {}
+
+#[derive(Clone, Copy, Default)]
+pub struct KeySpace<M: Mode> {
+    id: KeySpaceId,
+    _phantom: PhantomData<M>,
+}
+
+impl<M: Mode> From<KeySpace<M>> for Prefix {
+    fn from(s: KeySpace<M>) -> Self {
+        [M::PREFIX, s.id[0], s.id[1], s.id[2]]
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct ApiV2<M: Mode> {
+    keyspace: KeySpace<M>,
+}
+
+impl<M: Mode> RequestCodec for ApiV2<M> {
+    fn encode_key(&self, mut key: Vec<u8>) -> Vec<u8> {
+        let mut encoded = Vec::with_capacity(key.len() + KEYSPACE_PREFIX_LEN);
+        let prefix: Prefix = self.keyspace.into();
+
+        encoded.extend_from_slice(&prefix);
+        encoded.append(&mut key);
+        encoded
+    }
+
+    fn decode_key(&self, key: &mut Vec<u8>) -> Result<()> {
+        let prefix: Prefix = self.keyspace.into();
```

Comment:
There is an allocation every time encode and decode is called. We can make Keyspace store a final prefix internally and get reference of Prefix instead.

Reply:
There is no heap allocation happening here since the `Prefix` is a type alias of `[u8; 4]`. The reference size in the stack should be the same as `Prefix`.